### PR TITLE
Calculate group prices based on attendee's prices

### PR DESCRIPTION
**Pull request by kitsuta**
Fixes https://github.com/magfest/ubersystem/issues/2108, with some caveats.

### CODE REVIEW

File: uber/config.py [hunk 1]
```diff
@@ -102,20 +102,16 @@ def get_oneday_price(self, dt):
...
 
-            if c.HARDCORE_OPTIMIZATIONS_ENABLED:
-                # WARNING: EXTREMELY AGGRESSIVE. Don't run the DB query that gets # of badges sold in order
-                # to lighten the server load / blocking time on the DB. THIS ****BREAKS**** BADGE PRICE INCREASES
-                # THAT ARE BASED ON THE NUMBER OF TICKETS SOLD.  Only turn this on if you know EXACTLY what
-                # you are doing, your server is having its face melted off, and you have no other options.
-                # YOU HAVE BEEN WARNED!!!! -Dom
-                badges_sold = 0
-            else:
+            if not dt or c.HARDCORE_OPTIMIZATIONS_ENABLED:
```
Comment (kitsuta) [inline uber/config.py]: This now ignores badges_sold any time we pass a datetime variable in. I've changed the calls to never pass it in unless the attendee has a `registered` date (e.g., is not new). The warning about hardcore_optimizations_enabled is already in configspec.ini so I ended up deleting it from here because it was confusing otherwise.

File: uber/config.py [hunk 2]
```diff
@@ -396,7 +392,10 @@ def _unrepr(d):
 c.PRICE_LIMITS = {}
 for _opt, _val in c.BADGE_PRICES['attendee'].items():
     try:
-        date = c.EVENT_TIMEZONE.localize(datetime.strptime(_opt, '%Y-%m-%d'))
+        if ' ' in _opt:
```
Comment (kitsuta) [inline uber/config.py]: This allows us to specify an hour and minute for price bumps, which is required if we're going to set price bumps retroactively to match when our prices were raised by the badge cap.

File: uber/models.py [hunk 3]
```diff
@@ -946,14 +946,14 @@ def table_cost(self):
...
-        return c.DEALER_BADGE_PRICE if self.is_dealer else c.get_group_price(sa.localized_now())
+        return c.DEALER_BADGE_PRICE if self.is_dealer else c.get_group_price()
 
     @cost_property
     def badge_cost(self):
         total = 0
         for attendee in self.attendees:
             if attendee.paid == c.PAID_BY_GROUP:
-                total += c.DEALER_BADGE_PRICE if attendee.is_dealer else c.get_group_price(attendee.registered)
+                total += c.DEALER_BADGE_PRICE if attendee.is_dealer else max(0, attendee.badge_cost - c.GROUP_DISCOUNT)
```
Comment (kitsuta) [inline uber/models.py]: This is the major change of this PR - instead of calculating their badge cost every time, groups will now look to their attendees' badge costs. This allows us to set `overridden_price` on each individual badge, preserving its cost in spite of price bumps based on bucket-based pricing. It should also allow us to do things like have minors in groups.

File: uber/models.py [hunk 4]
```diff
@@ -1195,7 +1195,7 @@ def badge_cost(self):
             return c.get_attendee_price(registered)
 
     @cost_property
-    def discount(self):
+    def age_discount(self):
```
Comment (kitsuta) [inline uber/models.py]: This is an unrelated change - `discount` is way too generic a name for a property in the core system!

File: uber/models.py [hunk 5]
```diff
@@ -1215,7 +1215,11 @@ def total_cost(self):
 
     @property
     def amount_unpaid(self):
-        return max(0, self.total_cost - self.amount_paid)
+        if self.paid == c.PAID_BY_GROUP:
```
Comment (kitsuta) [inline uber/models.py]: This is the other part of the change that makes things work - we want to make sure attendees aren't charged for their badges if they're part of a group. The max(0, X) is used because some group's badges may be free.

**Comment by kitsuta** [comment]
This PR aims to change group functionality in kind of a big way, but amazingly enough I think most things will be unaffected. It's late though, so I may be wrong.

The rub here is that for groups currently in the system, we will have an incredibly hard time calculating their attendees' badge costs. My idea is:
1. Pull a list of groups, their current badge_cost, their registered date, and the # of "paid by group" badges they have. This will require a somewhat complex DB call.
2. Using Excel, divide all groups' badge costs by their "paid for" badges. Sort these by registered date.
3. Set the 'price bumps' datetimes to retroactively match when price bumps actually occurred for groups (Note: I tried to do this by looking at the FODC for attendee data, but it's not at all clearcut data - the only thing these price bumps will affect is groups, so that's all we should look at).
4. Merge the config change for that, plus this PR and the one for magprime. Resave all attendees so that registered attendees in groups get their `overridden_price` set to the correct price.
5. Find any groups whose badges do not conform to this (i.e., the cost per badge for them is a fraction) and fix their badges' costs manually.
6. Turn autorecalc on for any group whose badge_cost matches what their badges actually cost (OR another method where we can find all 'attendee' groups who should have autorecalc on)
7. Fix the code so that new badges only add to the group's cost if the group has autorecalc on.

File: uber/config.py (hunk 1 above)
Comment (binary1230) [inline uber/config.py]: I wouldn't mind keeping the warning here just because enabling c.HARDCORE_OPTIMIZATIONS is such a huge deal and I could see someone in the future enabling it without checking configspec.ini.  Sliiiigthly indifferent though.

Comment (binary1230) [inline uber/config.py]: unless I'm reading this wrong, the logic here for c.HARDCORE_OPTIMIZATIONS is now reversed.  i.e. when set, it should say badges_sold=0.  instead, it does badges_sold=self.BADGES_SOLD

File: uber/config.py (hunk 2 above)
Comment (binary1230) [inline uber/config.py]: nice!

File: uber/models.py (hunk 5 above)
Comment (binary1230) [inline uber/models.py]: this is a little OMHBITY, man these cost properties are so tough to keep track of mentally when reading the code at a glance.

File: uber/config.py [hunk 6]
```diff
@@ -102,18 +102,16 @@ def get_oneday_price(self, dt):
...
         price = self.INITIAL_ATTENDEE
         if self.PRICE_BUMPS_ENABLED:
 
-            if c.HARDCORE_OPTIMIZATIONS_ENABLED:
-                # WARNING: EXTREMELY AGGRESSIVE. Don't run the DB query that gets # of badges sold in order
-                # to lighten the server load / blocking time on the DB. THIS ****BREAKS**** BADGE PRICE INCREASES
-                # THAT ARE BASED ON THE NUMBER OF TICKETS SOLD.  Only turn this on if you know EXACTLY what
-                # you are doing, your server is having its face melted off, and you have no other options.
-                # YOU HAVE BEEN WARNED!!!! -Dom
+            if dt or c.HARDCORE_OPTIMIZATIONS_ENABLED:
```
Comment (binary1230) [inline uber/config.py]: +1 on this, looks good now

**Comment by kitsuta** [comment]
I got the data I needed, will be doing some number crunching tomorrow and fixing of config before merging this.

**Comment by kitsuta** [comment]
Final update: It took a view "oops, it's broken on staging" pull requests, but this actually seems to ahve worked. Amazing.
